### PR TITLE
Add fusion detection catalog and MLIR fusion-enablement skill

### PR DESCRIPTION
**Pull request by ctr-lelanchelianTT**
### Problem description
No systematic way to find a model's missed op fusions or tell whether each needs a flag, a pattern tweak, a new MLIR pattern, or a kernel.

### What's changed
- `finding-missed-fusions` (improved): audits TTIR/TTNN, writes `fusion_todo.yml`, classifies each candidate (`dialect_level` / `pass_status` / `owning_pass` / `existing_pattern`).
- `ttmlir-fusing-catalog.md` (new): source of truth for fusions tt-mlir already does — powers the classification.
- `enabling-fusions` (new): turns an actionable entry into a real tt-mlir pattern (scaffold, register, FileCheck, re-verify).

### Validation
Exercised end-to-end on YOLOv4: found the where-form LeakyReLU the old matcher missed, enabled it (70 `ttnn.leaky_relu` fired), **+37% throughput** (13.82 → 18.97 samples/s); correctly punted conv+mish/leaky_relu to kernel requests.

### Checklist
- [x] New/Existing tests provide coverage for changes

I have attached the logs for your reference:
[yolov4_after_clean.log](https://github.com/user-attachments/files/28451880/yolov4_after_clean.log)
[yolov4_before_clean.log](https://github.com/user-attachments/files/28451881/yolov4_before_clean.log)

### CODE REVIEW

**Comment by ctr-lelanchelianTT** [comment]
## Update: end-to-end validation of the fusion workflow on YOLOv4

The skills in this PR were exercised end-to-end on a real model. The prototype LeakyReLU
fusion that the PR note said *"isn't built/validated yet"* is now **built, FileCheck-gated,
verified-fired on the model, and benchmarked**. Summary below (the tt-mlir pattern code itself
still lives in the tt-mlir repo and is not part of this docs/skills PR).

### Phase 1 — LeakyReLU fusion (`finding-missed-fusions` → `enabling-fusions`)

**Root cause the audit surfaced:** YOLOv4's neck/head does **not** emit LeakyReLU as
`maximum(x, mul(x, α))`. It emits the **select/where form**
`where(gt(x, 0), x, multiply(x, 0.1))`, with `0` and `0.1` as splat **`ttir.constant`** ops
(not `ttir.full`). The backbone is **Mish** (already fused to `ttnn.mish`). The original
maximum-form matcher therefore produced **0** fused ops on YOLOv4.

**Fix:** added `LeakyReluWhereFusionPattern` (anchored on `WhereOp`) + a splat-aware
`getConstantOrFullScalar` helper to `TTIRFusing.cpp` (+190 lines), plus where-form positive/
negative cases in `leaky_relu_fusing.mlir`. Conservative: gt-against-zero, exact-input match,
single-use ops, `0 < α < 1` only.

**Verification:**
- FileCheck gate passes — 5 positive cases fuse, all negatives preserved.
- Fires on the real model: **70 `ttnn.leaky_relu`** in YOLOv4's TTNN IR (was 0).
- Correct by construction: `where(x>0, x, 0.1·x) ≡ leaky_relu(x, 0.1)`.

**Clean benchmark** (both runs no-debug, `opt_level=1, bs=1, fp32, 480×640`; baseline rebuilt
with the pattern *disabled* rather than reusing the debug-on log, since `TTXLA_LOGGER_LEVEL=DEBUG`
adds per-iter overhead):

| build | throughput | avg latency |
|---|---|---|
| before (fusion off) | 13.82 samples/s | 72.37 ms |
| **after (fusion on)** | **18.97 samples/s** | **52.72 ms** |
| **delta** | **+37%** | **−27%** |

```
# yolov4_before_clean.log  (fusion OFF)
opt_level=1 loop_count=32 warmup=3 batch_size=1 dtype=torch.float32
avg latency:      72.369 ms/iter
throughput:       13.82 samples/second

# yolov4_after_clean.log   (fusion ON, where-form pattern)
opt_level=1 loop_count=32 warmup=3 batch_size=1 dtype=torch.float32
avg latency:      52.720 ms/iter
throughput:       18.97 samples/second

# fired-count check (DEBUG IR dump)
$ grep -c "ttnn.leaky_relu" yolov4_after.log
70
```

### Phase 2 — next-fusion audit (`finding-missed-fusions` + catalog classification)

Produced `fusion_todo.yml` + `kernel_requests.md`. Key results:

- **conv + BatchNorm and conv + bias are already handled** — zero `ttnn.batch_norm` in the
  final TTNN; folded into the 110 `prepare_conv2d_weights`/`prepare_conv2d_bias` at const-eval.
  (So `conv2dWithMultiplyEnabled` would be redundant here — verified, not assumed.)
- The only remaining high-count sequences are **conv2d + mish (72)** and
  **conv2d + leaky_relu (35)** — both conv-fused-activation folds. Neither is a Tier-1
  MLIR-only win:
  - `Mish` is not in the TTNN `UnaryOpType` enum and `Conv2dConfig.activation` is a
    parameterless `UnaryOpType` → can't be expressed; no tt-metal mish-in-conv kernel.
  - `LeakyRelu` *is* a `UnaryOpType`, but conv's activation can't carry the `0.1` slope
    (only matmul's `UnaryWithParamAttr` carries params), and tt-metal conv leaky_relu
    support is unconfirmed.
- Both recorded in `kernel_requests.md` as tt-metal capability requests — deliberately
  **not** implemented (Tier-2 / kernel-dependent), per the skill's guardrails.

**Net:** the `finding-missed-fusions` → catalog-classification → `enabling-fusions` loop
worked as designed: it found a real missed fusion (where-form LeakyReLU), guided a conservative
pattern, the result fired and gave a measured **+37%**, and it correctly classified the
remaining candidates as kernel work rather than producing a perf-neutral or incorrect fusion.

**Review by meenakshiramanathan1 — APPROVED** [review]

**Review by saiarthiraguram — APPROVED** [review]
